enhancements/update/automatic-updates.md

```diff
@@ -0,0 +1,176 @@
+authors:
+  - "@wking"
+reviewers:
+  - "@abhinavdahiya"
+  - "@crawford"
+  - "@smarterclayton"
+approvers:
+  - TBD
+creation-date: 2019-11-19
+last-updated: 2019-11-21
+status: implementable
+---
+
+# Automatic Updates
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Operator-managed updates are [one of the key benefits of OpenShift 4][architecture].
+Cluster update suggestions are distributed via [Cincinnati][cincinnati-openshift], so clusters only receive update recommendations that are appropriate for their cluster.
+While manually approving updates at a per-cluster level is possible, it should not be required.
+This enhancement adds a new `automaticUpdates` property to [the ClusterVersion `spec`][api-spec] so cluster administrators can opt in to or out of automatic updates.
+
+## Motivation
+
+Manually approving updates is tedious, doesn't scale well for administrators managing many clusters, and delays the application of potential security fixes.
+Administrators may wish to enable automatic updates for any of those or other reasons.
+Currently, administrators must implement custom polling logic to check for and apply any available updates.
+This enhancement would provide those administrators with a convenient property instead.
+
+### Goals
+
+* Make it easy to opt in to and out of automatic updates.
+
+### Non-Goals
+
+* Support calendar gating or other logic about when updates may be attempted.
+* Choose whether the installer defaults to enabling automatic updates or not for new clusters.
+
+## Proposal
+
+As proposed in [this API pull request][api-pull-request], to add a new property:
+
+```go
+// automaticUpdates enables automatic updates.
+// +optional
+AutomaticUpdates bool `json:"automaticUpdates,omitempty"`
+```
+
+to [`ClusterVersionSpec`][api-spec] to record the administrator's preference.
+
+### User Stories
+
+#### Leading Clusters
+
+Alice has clusters she wants to update as soon as an updated release is available in her configured channel.
+Before this enhancement she would need to construct a poller to run `oc adm upgrade --to-latest` or similar.
+With this enhancement, she can set `automaticUpdates` and does not need to write an external poller.
+
+### Risks and Mitigations
+
+Cluster-managed updates are a relatively new thing, and while we have dozens of OpenShift 4 releases out so far, we have had a few updates result in stuck clusters (like [this][machine-config-operator-drain-wedge].
+Allowing automatic updates would make it more likely that update failures happen when there is no administrator actively watching to notice and recover from the failure.
+This is mitigated by:
+
+* Alerting.
```

Comment:
We actually know that certain (non critical) alerts will fire during upgrades. The observability group (of which the in-cluster monitoring team is a part of) has been trialing silencing everything but critical alerts during upgrades. While this is not perfect yet, I think this is what we ultimately want to automate for automatic upgrades.

That said any `severity: critical` alerts firing should prevent an automatic upgrade from happening. This level is used to page people, so in a production cluster this should already mean that someone has been notified. Our goal must be that no critical alerts are firing during upgrades. Essentially this means we need to set certain objectives per component and if they are violated we cannot proceed with an automatic upgrade and/or people are notified. I know the in-cluster team is planning an initiative to introduce these objectives for components (at this point earliest in the 4.5 time frame I would say).

Reply:
> We actually know that certain (non critical) alerts will fire during upgrades.

Do we consider these overly sensitive alerts?  Update bugs?  I'd expect that our goal is to have updates be smooth enough that we can apply them without additional alerts firing.  Silencing non-critical alerts sounds like a reasonable stopgap to avoid alert fatigue, but I'm less comfortable with it as a long-term plan.

> That said any severity: critical alerts firing should prevent an automatic upgrade from happening.

Hmm.  We might be able to swing this with the monitoring operator setting [`Upgradeable=False`][1] when it sees a critical alert.

[1]: https://github.com/openshift/api/blob/c562e326a6ae5d8d7167cf9787684098cf03190a/config/v1/types_cluster_operator.go#L170-L174

Reply:
Silencing is made exactly for the case of "we know a maintenance window is coming, so don't bother us with non critical things", so I think it's exactly the right thing. Warning alerts are more the type of "something might be up, but it's not urgent". They should typically just open a ticket for someone to eventually take a look at. Critical alerts are the type of alerts that tell someone that there is something that needs attention urgently and that SLOs are likely to be violated.

> We might be able to swing this with the monitoring operator setting Upgradeable=False when it sees a critical alert.

I think that's a good idea.